src/sentry/api/endpoints/project_codeowners.py

```diff
@@ -1,11 +1,146 @@
+    ProjectCodeOwners,
+    ExternalUser,
+    ExternalTeam,
+    RepositoryProjectPathConfig,
+    UserEmail,
+)
+from sentry.api.serializers.rest_framework.base import CamelSnakeModelSerializer
+from sentry.ownership.grammar import (
+    parse_code_owners,
+    convert_codeowners_syntax,
+)
+
+from sentry.api.endpoints.project_ownership import ProjectOwnershipSerializer, ProjectOwnershipMixin
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectCodeOwnerSerializer(CamelSnakeModelSerializer):
+    code_mapping_id = serializers.IntegerField(required=True)
+    raw = serializers.CharField(required=True)
+    organization_integration_id = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = ProjectCodeOwners
+        fields = ["raw", "code_mapping_id", "organization_integration_id"]
+
+    def validate(self, attrs):
+        # If it already exists, set default attrs with existing values
+        if self.instance:
+            attrs = {
+                "raw": self.instance.raw,
+                "code_mapping_id": self.instance.repository_project_path_config,
+                **attrs,
+            }
+
+        if not attrs.get("raw", "").strip():
+            return attrs
+        external_association_err = ""
+        # Get list of team/user names from CODEOWNERS file
+        teamnames, usernames, emails = parse_code_owners(attrs["raw"])
+
+        # Check if there exists Sentry users with the emails listed in CODEOWNERS
+        user_emails = UserEmail.objects.filter(email__in=emails)
+        user_emails = [user.email for user in user_emails]
+        user_emails_diff = [email for email in emails if email not in user_emails]
+
+        if len(user_emails_diff):
+            external_association_err += f'The following emails do not have an user associated in Sentry: {", ".join(user_emails_diff)}.'
+
+        # Check if the usernames have an association
+        external_users = ExternalUser.objects.filter(external_name__in=usernames)
+
+        external_users_names = [user.external_name for user in external_users]
+        external_users_diff = [name for name in usernames if name not in external_users_names]
+
+        if len(external_users_diff):
+            nl = "\n" if len(external_association_err) else ""
+            external_association_err += f'{nl}The following usernames do not have an association in Sentry: {", ".join(external_users_diff)}.'
+
+        # Check if the team names have an association
+        external_teams = ExternalTeam.objects.filter(external_name__in=teamnames)
+
+        external_teams_names = [team.external_name for team in external_teams]
+        external_teams_diff = [name for name in teamnames if name not in external_teams_names]
 
+        if len(external_teams_diff):
+            nl = "\n" if len(external_association_err) else ""
+            external_association_err += f'{nl}The following team names do not have an association in Sentry: {", ".join(external_teams_diff)}.'
```

Comment:
@NisanthanNanthakumar Nit: I would recommend just making an array of errors then calling `"\n".join` on the array at the end

Reply:
@scefali 👍 smart!